contracts/src/factory/HyperdriveRegistry.sol

```diff
@@ -1,62 +1,392 @@
+            }
+            // If the updated data is non-zero and the existing data is zero,
+            // we are adding a new entry to the registry.
+            else if (_data[i] != 0 && data == 0) {
+                _addInstance(__instances[i], _data[i], __factories[i]);
+            }
+            // Otherwise, the update is attempting to remove a non-existant
+            // entry. To avoid emitting a confusing event, we continue.
+            else {
+                continue;
+            }
+
+            // Emit an event recording the update.
+            emit InstanceInfoUpdated(__instances[i], _data[i], __factories[i]);
+        }
+    }
+
+    /// @inheritdoc IHyperdriveRegistry
+    function getNumberOfFactories() external view returns (uint256) {
+        return _factories.length;
+    }
+
+    /// @inheritdoc IHyperdriveRegistry
+    function getFactoriesInRange(
+        uint256 _startIndex,
+        uint256 _endIndex
+    ) external view returns (address[] memory factories) {
+        // If the indexes are malformed, revert.
+        if (_startIndex > _endIndex) {
+            revert IHyperdriveGovernedRegistry.InvalidIndexes();
+        }
+        if (_endIndex > _factories.length) {
+            revert IHyperdriveGovernedRegistry.EndIndexTooLarge();
+        }
+
+        // Get the registered factories in the range.
+        factories = new address[](_endIndex - _startIndex);
+        for (uint256 i = _startIndex; i < _endIndex; i++) {
+            factories[i - _startIndex] = _factories[i];
+        }
+
+        return factories;
+    }
+
+    /// @inheritdoc IHyperdriveRegistry
+    function getFactoryAtIndex(uint256 _index) external view returns (address) {
+        return _factories[_index];
+    }
+
+    /// @inheritdoc IHyperdriveRegistry
+    function getFactoryInfo(
+        address[] calldata __factories
+    ) external view override returns (FactoryInfo[] memory info) {
+        info = new FactoryInfo[](__factories.length);
+        for (uint256 i = 0; i < __factories.length; i++) {
+            IHyperdriveFactory factory = IHyperdriveFactory(__factories[i]);
+            info[i] = FactoryInfo({
+                data: _factoryInfo[__factories[i]].data,
+                name: factory.name(),
+                version: factory.version()
+            });
+        }
+        return info;
+    }
+
+    /// @inheritdoc IHyperdriveRegistry
+    function getNumberOfInstances() external view returns (uint256) {
+        return _instances.length;
     }
 
     /// @inheritdoc IHyperdriveRegistry
-    function getHyperdriveInfo(
-        address _hyperdriveInstance
-    ) external view override returns (uint256) {
-        return _hyperdriveInfo[_hyperdriveInstance];
+    function getInstancesInRange(
+        uint256 _startIndex,
+        uint256 _endIndex
+    ) external view returns (address[] memory instances) {
+        // If the indexes are malformed, revert.
+        if (_startIndex > _endIndex) {
```

Comment:
what if _startIndex == _endIndex? I think we should fail in this case bc it won't return anything, right?